### PR TITLE
DSLParser: pass `function` as a parameter to all `state_foo` methods

### PR DESCRIPTION
**Pull request by AlexWaygood**
Towards #7

### CODE REVIEW

**Comment by AlexWaygood** [comment]
It looks like a big diff, but a lot of it is just me refusing to keep to the existing ~120-character line lengths for bits that I have to touch anyway with this refactor 😅

This _doesn't_ get rid of all the `assert isinstance(self.function, Function)` assertions. But it does get rid of some. And, more importantly, it means that `function` is now _always_ passed as a parameter to all methods on the `DSLParser`. This feels much easier to understand to me than the current situation, where some methods on `DSLParser` access it via `self.function`, and others have it passed as a parameter.

Ultimately, if it's sometimes `None` and sometimes an instance of `Function`, it just doesn't feels suitable to keep it as state on `DSLParser` instances.

File: Tools/clinic/clinic.py [hunk 1]
```diff
@@ -4688,8 +4697,10 @@ def state_modulename_name(self, line: str) -> None:
         # this line is permitted to start with whitespace.
         # we'll call this number of spaces F (for "function").
 
+        assert function is None
+
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py:4701]: This assertion (and a few others I add in this PR) isn't actually necessary to keep mypy happy, but I think it's useful to assert exactly where we are in the state machine at this point. I.e., it's there as a sanity check and for readability.

**Comment by AlexWaygood** [comment]
Note: I haven't checked test coverage for all the lines I'm touching in this PR. Will do that before filing a PR to CPython -- if you like the overall idea here!

File: Tools/clinic/clinic.py [hunk 2]
```diff
@@ -4860,33 +4878,37 @@ def state_modulename_name(self, line: str) -> None:
...
 
-    def state_parameters_start(self, line: str) -> None:
-        if not self.valid_line(line):
-            return
-
-        # if this line is not indented, we have no parameters
-        if not self.indent.infer(line):
-            return self.next(self.state_function_docstring, line)
+    def state_parameters_start(self, function: Function | None, line: str) -> Function:
+        assert function is not None
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py:4882]: And another

**Comment by erlend-aasland** [comment]
I'll have a look later! Are you thinking of ripping out the state machine in the long term?

**Comment by AlexWaygood** [comment]
> Are you thinking of ripping out the state machine in the long term?

Not sure... I do feel like the current way of calling all the `state_foo` methods via `self.next()` is a bit of a leaky abstraction (the `state_foo` methods just have too many differences between them!), and this problem is sort-of a symptom of that.

But this change definitely makes me feel happier about the overall design. So I think it's a good step _for now_ :)

**Pull request closed without merging**